www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,230 @@
+
+![Commit prepared transaction flow](/blog/2024-11-19-distributed-ddl/prepare_xact.png)
+
+Once all the databases commit, the second phase of the 2PC is completed, and the distributed transaction is fully committed. We still need to discuss what happens if they don't commit successfully. But before diving into that, lets take a quick detour and look at how we handle locks.
+
+## Locks
+
+If you recall back, the first step the `nile_ddl` takes toward distributing a DDL, is to determine the locks the DDL will need and attempt to acquire them on all relevant databases. We do this with 3 simple goals in mind:
+
+1. **Fail early:** If a DDL is going to fail due to inability to acquire locks, we want to know about this before performing any resource intensive operations.
+2. **Don't block for too long:** Many DDLs require `ACCESS EXCLUSIVE` lock. Attempts to acquire this lock will block other sessions, even while waiting for the lock to get acquired. On a busy OLTP system it is important to minimize such blocks.
+3. **Don't deadlock:** We want to avoid a situation where two sessions are stuck, each waiting for the other to release a lock and both failing to make progress. 
+
+In order to fail early, we make sure to acquire all the necessary locks as the first step in processing a DDL. Before we even run the DDL locally. We could do it in parallel, but this can be very wasteful - imagine building an index on a large table locally, only to roll it back because we couldn't acquire the lock on a table on one of the remote databases. 
+
+So the `nile_ddl` extension takes the DDL and tries to determine which locks it will need. `ALTER` and `DROP` are fairly straight forward, we need to acquire a lock on the objects that the DDL is modifying. In most cases, acquiring the lock is simply a matter of sending `LOCK TABLE <table name> IN <lock level> MODE` with the same level that the DDL will require later. So before distributing `DROP TABLE todos`, the transaction coordinator must distribute `LOCK TABLE todos ACCESS EXLUSIVE MODE`. However, `LOCK TABLE` only applies to... well, tables. And we sometimes need to lock other objects. For example `DROP SCHEMA` requires locking the schema. Postgres has internal methods for locking other objects, so in our extension, we added the following logic:
```

Comment:
drop is straightforward unless you do DROP CASCADE, as you need to recurse through the dependent objects.